dnscache/storage.go

```diff
@@ -0,0 +1,94 @@
+package dnscache
+
+import (
+	"net"
+	"reflect"
+	"time"
+
+	"fmt"
+
+	"github.com/pmylund/go-cache"
+)
+
+// DnsCacheItem represents single record in cache
+type DnsCacheItem struct {
+	addrs []string
+}
+
+func (item *DnsCacheItem) IsEqualsTo(addrs []string) bool {
+	return reflect.DeepEqual(item.addrs, addrs)
+}
+
+// DnsCacheStorage is an in-memory cache of auto-purged dns query ip responses
+type DnsCacheStorage struct {
+	cache *cache.Cache
+}
+
+func NewDnsCacheStorage(expiration, checkInterval time.Duration) *DnsCacheStorage {
+	storage := DnsCacheStorage{cache.New(expiration, checkInterval)}
+	return &storage
+}
+
+// Items returns map of non expired dns cache items
+func (dc *DnsCacheStorage) Items(includeExpired bool) map[string]DnsCacheItem {
+	var allItems = dc.cache.Items()
+
+	nonExpiredItems := map[string]DnsCacheItem{}
+
+	for k, v := range allItems {
+		if !includeExpired && v.Expired() {
+			continue
+		}
+		nonExpiredItems[k] = v.Object.(DnsCacheItem)
+	}
+
+	return nonExpiredItems
+}
+
+// Get returns non expired item from cache
+func (dc *DnsCacheStorage) Get(key string) (DnsCacheItem, bool) {
+	item, found := dc.cache.Get(key)
+	if !found {
+		return DnsCacheItem{}, false
+	}
+	return item.(DnsCacheItem), found
+}
+
+func (dc *DnsCacheStorage) Delete(key string) {
+	dc.cache.Delete(key)
+}
+
+// FetchItem returns list of ips from cache or resolves them and add to cache
+func (dc *DnsCacheStorage) FetchItem(hostName string) ([]string, error) {
+	if hostName == "" {
+		return nil, fmt.Errorf("hostName can't be empty. hostName=%v", hostName)
+	}
+
+	item, ok := dc.Get(hostName)
+	if ok {
+		logger.Debugf("Dns record was populated from cache: hostName=%q, addrs=%q", hostName, item.addrs)
```

Comment:
following logrus-convention it might better to do (so debug message context will appear as message fields):

```
logger.WithFields(logrus.Fields{
  "hostName": hostName,
  "addrs": item.addrs,
}.Debug("Dns record was populated from cache")
```

Reply:
fixed in 7f332e6a861e73fc8eb846e0e9aadd5f96fc724e